File: rust/src/resource/set.rs
use tonic::{Request, Status};
use uuid::Uuid;
use rmcs_resource_api::set::set_service_client::SetServiceClient;
use rmcs_resource_api::set::{
    SetSchema, SetId, SetIds, SetName, SetUpdate, SetMemberRequest, SetMemberSwap,
    SetTemplateSchema, SetTemplateId, SetTemplateIds, SetTemplateName, SetTemplateUpdate, SetTemplateMemberRequest, SetTemplateMemberSwap
};
use crate::resource::Resource;
use rmcs_api_server::utility::interceptor::TokenInterceptor;

const SET_NOT_FOUND: &str = "requested set not found";

pub(crate) async fn read_set(resource: &Resource, id: Uuid)
    -> Result<SetSchema, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetId {
        id: id.as_bytes().to_vec()
    });
    let response = client.read_set(request)
        .await?
        .into_inner();
    Ok(response.result.ok_or(Status::not_found(SET_NOT_FOUND))?)
}

pub(crate) async fn list_set_by_ids(resource: &Resource, ids: &[Uuid])
    -> Result<Vec<SetSchema>, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetIds {
        ids: ids.into_iter().map(|&id| id.as_bytes().to_vec()).collect()
    });
    let response = client.list_set_by_ids(request)
        .await?
        .into_inner();
    Ok(response.results)
}

pub(crate) async fn list_set_by_template(resource: &Resource, template_id: Uuid)
    -> Result<Vec<SetSchema>, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateId {
        id: template_id.as_bytes().to_vec()
    });
    let response = client.list_set_by_template(request)
        .await?
        .into_inner();
    Ok(response.results)
}

pub(crate) async fn list_set_by_name(resource: &Resource, name: &str)
    -> Result<Vec<SetSchema>, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetName {
        name: name.to_owned()
    });
    let response = client.list_set_by_name(request)
        .await?
        .into_inner();
    Ok(response.results)
}

pub(crate) async fn create_set(resource: &Resource, id: Uuid, template_id: Uuid, name: &str, description: Option<&str>)
    -> Result<Uuid, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetSchema {
        id: id.as_bytes().to_vec(),
        template_id: template_id.as_bytes().to_vec(),
        name: name.to_owned(),
        description: description.unwrap_or_default().to_owned(),
        members: Vec::new()
    });
    let response = client.create_set(request)
        .await?
        .into_inner();
    Ok(Uuid::from_slice(&response.id).unwrap_or_default())
}

pub(crate) async fn update_set(resource: &Resource, id: Uuid, template_id: Option<Uuid>, name: Option<&str>, description: Option<&str>)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetUpdate {
        id: id.as_bytes().to_vec(),
        template_id: template_id.map(|id| id.as_bytes().to_vec()),
        name: name.map(|s| s.to_owned()),
        description: description.map(|s| s.to_owned())
    });
    client.update_set(request)
        .await?;
    Ok(())
}

pub(crate) async fn delete_set(resource: &Resource, id: Uuid)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetId {
        id: id.as_bytes().to_vec()
    });
    client.delete_set(request)
        .await?;
    Ok(())
}

pub(crate) async fn add_set_member(resource: &Resource, id: Uuid, device_id: Uuid, model_id: Uuid, data_index: &[u8])
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetMemberRequest {
        set_id: id.as_bytes().to_vec(),
        device_id: device_id.as_bytes().to_vec(),
        model_id: model_id.as_bytes().to_vec(),
        data_index: data_index.to_owned()
    });
    client.add_set_member(request)
        .await?;
    Ok(())
}

pub(crate) async fn remove_set_member(resource: &Resource, id: Uuid, device_id: Uuid, model_id: Uuid)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetMemberRequest {
        set_id: id.as_bytes().to_vec(),
        device_id: device_id.as_bytes().to_vec(),
        model_id: model_id.as_bytes().to_vec(),
        data_index: Vec::new()
    });
    client.remove_set_member(request)
        .await?;
    Ok(())
}

pub(crate) async fn swap_set_member(resource: &Resource, id: Uuid, device_id_1: Uuid, model_id_1: Uuid, device_id_2: Uuid, model_id_2: Uuid)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetMemberSwap {
        set_id: id.as_bytes().to_vec(),
        device_id_1: device_id_1.as_bytes().to_vec(),
        model_id_1: model_id_1.as_bytes().to_vec(),
        device_id_2: device_id_2.as_bytes().to_vec(),
        model_id_2: model_id_2.as_bytes().to_vec(),
    });
    client.swap_set_member(request)
        .await?;
    Ok(())
}

pub(crate) async fn read_set_template(resource: &Resource, id: Uuid)
    -> Result<SetTemplateSchema, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateId {
        id: id.as_bytes().to_vec()
    });
    let response = client.read_set_template(request)
        .await?
        .into_inner();
    Ok(response.result.ok_or(Status::not_found(SET_NOT_FOUND))?)
}

pub(crate) async fn list_set_template_by_ids(resource: &Resource, ids: &[Uuid])
    -> Result<Vec<SetTemplateSchema>, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateIds {
        ids: ids.into_iter().map(|&id| id.as_bytes().to_vec()).collect()
    });
    let response = client.list_set_template_by_ids(request)
        .await?
        .into_inner();
    Ok(response.results)
}

pub(crate) async fn list_set_template_by_name(resource: &Resource, name: &str)
    -> Result<Vec<SetTemplateSchema>, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateName {
        name: name.to_owned()
    });
    let response = client.list_set_template_by_name(request)
        .await?
        .into_inner();
    Ok(response.results)
}

pub(crate) async fn create_set_template(resource: &Resource, id: Uuid, name: &str, description: Option<&str>)
    -> Result<Uuid, Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateSchema {
        id: id.as_bytes().to_vec(),
        name: name.to_owned(),
        description: description.unwrap_or_default().to_owned(),
        members: Vec::new()
    });
    let response = client.create_set_template(request)
        .await?
        .into_inner();
    Ok(Uuid::from_slice(&response.id).unwrap_or_default())
}

pub(crate) async fn update_set_template(resource: &Resource, id: Uuid, name: Option<&str>, description: Option<&str>)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateUpdate {
        id: id.as_bytes().to_vec(),
        name: name.map(|s| s.to_owned()),
        description: description.map(|s| s.to_owned())
    });
    client.update_set_template(request)
        .await?;
    Ok(())
}

pub(crate) async fn delete_set_template(resource: &Resource, id: Uuid)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateId {
        id: id.as_bytes().to_vec()
    });
    client.delete_set_template(request)
        .await?;
    Ok(())
}

pub(crate) async fn add_set_template_member(resource: &Resource, id: Uuid, type_id: Uuid, model_id: Uuid, data_index: &[u8])
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateMemberRequest {
        set_id: id.as_bytes().to_vec(),
        type_id: type_id.as_bytes().to_vec(),
        model_id: model_id.as_bytes().to_vec(),
        data_index: data_index.to_owned(),
        template_index: 0
    });
    client.add_set_template_member(request)
        .await?;
    Ok(())
}

pub(crate) async fn remove_set_template_member(resource: &Resource, id: Uuid, index: usize)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateMemberRequest {
        set_id: id.as_bytes().to_vec(),
        type_id: Uuid::nil().as_bytes().to_vec(),
        model_id: Uuid::nil().as_bytes().to_vec(),
        data_index: Vec::new(),
        template_index: index as i32
    });
    client.remove_set_template_member(request)
        .await?;
    Ok(())
}

pub(crate) async fn swap_set_template_member(resource: &Resource, id: Uuid, index_1: usize, index_2: usize)
    -> Result<(), Status>
{
    let interceptor = TokenInterceptor(resource.access_token.clone());
    let mut client = 
        SetServiceClient::with_interceptor(resource.channel.to_owned(), interceptor);
    let request = Request::new(SetTemplateMemberSwap {
        set_id: id.as_bytes().to_vec(),
        template_index_1: index_1 as i32,
        template_index_2: index_2 as i32
    });
    client.swap_set_template_member(request)
        .await?;
    Ok(())
}

File: python/src/rmcs_api_client/resource/set.py
from rmcs_resource_api import set_pb2, set_pb2_grpc
from typing import Optional, List
from dataclasses import dataclass
from uuid import UUID
import grpc


@dataclass
class SetMember:
    device_id: UUID
    model_id: UUID
    data_index: List[int]

    def from_response(r):
        return SetMember(UUID(bytes=r.device_id), UUID(bytes=r.model_id), list(r.data_index))

@dataclass
class SetSchema:
    id: UUID
    template_id: UUID
    name: str
    description: str
    members: List[SetMember]

    def from_response(r):
        members = []
        for member in r.members: members.append(SetMember.from_response(member))
        return SetSchema(UUID(bytes=r.id), UUID(bytes=r.template_id), r.name, r.description, members)

@dataclass
class SetTemplateMember:
    type_id: UUID
    model_id: UUID
    data_index: List[int]

    def from_response(r):
        return SetTemplateMember(UUID(bytes=r.type_id), UUID(bytes=r.model_id), list(r.data_index))

@dataclass
class SetTemplateSchema:
    id: UUID
    name: str
    description: str
    members: List[SetTemplateMember]

    def from_response(r):
        members = []
        for member in r.members: members.append(SetTemplateMember.from_response(member))
        return SetTemplateSchema(UUID(bytes=r.id), r.name, r.description, members)


def read_set(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetId(id=id.bytes)
        response = stub.ReadSet(request=request, metadata=resource.metadata)
        return SetSchema.from_response(response.result)

def list_set_by_ids(resource, ids: list[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetIds(ids=list(map(lambda x: x.bytes, ids)))
        response = stub.ListSetByIds(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def list_set_by_template(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateId(id=id.bytes)
        response = stub.ListSetByTemplate(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def list_set_by_name(resource, name: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetName(name=name)
        response = stub.ListSetByName(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def list_set_option(resource, template_id: Optional[UUID], name: Optional[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        template_bytes = None
        if template_id != None: template_bytes = template_id.bytes
        request = set_pb2.SetOption(
            template_id=template_bytes,
            name=name
        )
        response = stub.ListSetOption(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def create_set(resource, id: UUID, template_id: UUID, name: str, description: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetSchema(
            id=id.bytes,
            template_id=template_id.bytes,
            name=name,
            description=description
        )
        response = stub.CreateSet(request=request, metadata=resource.metadata)
        return UUID(bytes=response.id)

def update_set(resource, id: UUID, template_id: Optional[UUID], name: Optional[str], description: Optional[str]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        template_bytes = None
        if template_id != None: template_bytes = template_id.bytes
        request = set_pb2.SetUpdate(
            id=id.bytes,
            template_id=template_bytes,
            name=name,
            description=description
        )
        stub.UpdateSet(request=request, metadata=resource.metadata)

def delete_set(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetId(id=id.bytes)
        stub.DeleteSet(request=request, metadata=resource.metadata)

def add_set_member(resource, id: UUID, device_id: UUID, model_id: UUID, data_index: list[int]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetMemberRequest(
            set_id=id,
            device_id=device_id,
            model_id=model_id,
            data_index=bytes(data_index)
        )
        stub.AddSetMember(request=request, metadata=resource.metadata)

def remove_set_member(resource, id: UUID, device_id: UUID, model_id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetMemberRequest(
            set_id=id,
            device_id=device_id,
            model_id=model_id
        )
        stub.RemoveSetMember(request=request, metadata=resource.metadata)

def swap_set_member(resource, id: UUID, device_id_1: UUID, model_id_1: UUID, device_id_2: UUID, model_id_2: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetMemberSwap(
            set_id=id,
            device_id_1=device_id_1,
            model_id_1=model_id_1,
            device_id_2=device_id_2,
            model_id_2=model_id_2
        )
        stub.SwapSetMember(request=request, metadata=resource.metadata)

def read_set_template(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateId(id=id.bytes)
        response = stub.ReadSetTemplate(request=request, metadata=resource.metadata)
        return SetTemplateSchema.from_response(response.result)

def list_set_template_by_ids(resource, ids: list[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateIds(ids=list(map(lambda x: x.bytes, ids)))
        response = stub.ListSetTemplateByIds(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetTemplateSchema.from_response(result))
        return ls

def list_set_template_by_name(resource, name: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateName(name=name)
        response = stub.ListSetTemplateByName(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetTemplateSchema.from_response(result))
        return ls

def list_set_template_option(resource, name: Optional[UUID]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetOption(
            name=name
        )
        response = stub.ListSetTemplateOption(request=request, metadata=resource.metadata)
        ls = []
        for result in response.results: ls.append(SetSchema.from_response(result))
        return ls

def create_set_template(resource, id: UUID, name: str, description: str):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateSchema(
            id=id.bytes,
            name=name,
            description=description
        )
        response = stub.CreateSetTemplate(request=request, metadata=resource.metadata)
        return UUID(bytes=response.id)

def update_set_template(resource, id: UUID, name: Optional[str], description: Optional[str]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateUpdate(
            id=id.bytes,
            name=name,
            description=description
        )
        stub.UpdateSetTemplate(request=request, metadata=resource.metadata)

def delete_set_template(resource, id: UUID):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateId(id=id.bytes)
        stub.DeleteSetTemplate(request=request, metadata=resource.metadata)

def add_set_template_member(resource, id: UUID, type_id: UUID, model_id: UUID, data_index: list[int]):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateMemberRequest(
            set_id=id,
            type_id=type_id,
            model_id=model_id,
            data_index=bytes(data_index)
        )
        stub.AddSetTemplateMember(request=request, metadata=resource.metadata)

def remove_set_template_member(resource, id: UUID, index: int):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateMemberRequest(
            set_id=id,
            template_index=index
        )
        stub.RemoveSetTemplateMember(request=request, metadata=resource.metadata)

def swap_set_template_member(resource, id: UUID, index_1: int, index_2: int):
    with grpc.insecure_channel(resource.address) as channel:
        stub = set_pb2_grpc.SetServiceStub(channel)
        request = set_pb2.SetTemplateMemberSwap(
            set_id=id,
            template_index_1=index_1,
            template_index_2=index_2
        )
        stub.SwapSetTemplateMember(request=request, metadata=resource.metadata)
